### DISCUSSION TITLE
RuntimeError: Python -> C++ conversions

### DISCUSSION THREAD

**Discussion by bullette007** [mitsuba-renderer/mitsuba3#792] [Q&A]
Hi everyone,

I'm currently working on a project where I'd like to implement a custom BSDF plugin, so I started with the tutorial (https://mitsuba.readthedocs.io/en/stable/src/others/custom_plugin.html).

Unfortunately, even the unmodified sample (only modified by setting the variant to `scalar_rgb`) does fail while calling `mi.render` with the following output:
`RuntimeError: When called outside a bound function, py::cast() cannot do Python -> C++ conversions which require the creation of temporary values`

This seems to have broken during the transition from version 3.0.2 (here it still works) to 3.1.0.

Any thoughts / help on this?

Many thanks in advance!

**Comment by njroussel** [comment]
Hi @bullette007 

In general custom Python plugins are not meant to be used with `scalar-*` variants. As mentioned in the tutorial, there is a significant performance overhead with Python and the JIT enabled variants mitigate that overhead.

I'm unable to fully reproduce the behaviour you describe. My best guess is that the `eval_pdf` method is returning `return 0.0, 0.0` when it should maybe explicitly return `return Color3f(0.0), 0.0`. More generally, I'm actually surprised you got this too work at all, even in `3.02`. Are you running this code on a single-threaded machine?
